src/data_generators/tests/copula_data_set_test.py

```diff
@@ -130,6 +135,224 @@ def test_fully_negatively_correlated_copula(self):
+            res = CopulaCorrelationMatrixGenerator.random(
+                p=5, rng=np.random.default_rng(seed)
+            )
+            # For a randomly generated correlation matrix, we test if it's positive definite
+            self.assertTrue(all(np.linalg.eig(res)[0] > 0))
+
+    def test_random_uniformity_with_two_pubs(self):
+        rng = np.random.default_rng(0)
+        cor_mats = [
+            CopulaCorrelationMatrixGenerator.random(p=2, rng=rng) for _ in range(400)
+        ]
+        cors = [cor_mat[0, 1] for cor_mat in cor_mats]
+        correlation_frequencies = np.histogram(a=cors, bins=4, range=(-1, 1))[0] / 400
+        np.testing.assert_almost_equal(correlation_frequencies, [1 / 4] * 4, decimal=1)
+
+    def test_random_uniformity_with_three_pubs(self):
+        rng = np.random.default_rng(0)
+        cor_mats = [
+            CopulaCorrelationMatrixGenerator.random(p=3, rng=rng) for _ in range(1000)
+        ]
+
+        def to_polar_coordinates(
+            x: float, y: float, z: float
+        ) -> Tuple[float, float, float]:
+            """Find the polar coordinates of any 3D point.
+
+            Find (r, theta, phi) such that
+            - x = r cos(phi) sin(theta)
+            - y = r sin(phi) sin(theta)
+            - z = r cos(theta).
+            """
+            r = np.sqrt(x ** 2 + y ** 2 + z ** 2)
+            phi = np.arctan2(y, x)
+            theta = np.arctan2(np.sqrt(x ** 2 + y ** 2), z)
+            return (r, phi, theta)
+
+        correlation_tuples = [
+            (cor_mat[0, 1], cor_mat[0, 2], cor_mat[1, 2]) for cor_mat in cor_mats
+        ]
+        polar_tuples = [
+            to_polar_coordinates(*correlation_tuple)
+            for correlation_tuple in correlation_tuples
+        ]
+
+        # 1. Test the uniformity of the radius.  Because of the positive definitiveness
+        # constraint, not all values of the radius are possible.  It can be proved that
+        # When (x, y, z) has a radius no greater than 0.5, the correlation matrix of the
+        # correlations (x, y, z) is always positive definite.
+        # We then expect that the radius_cube = radius ** 3 of the (x, y, z) of
+        # the generated matrix is uniform at least in (0, 0.5).  This is mathematically
+        # proved to be a necessary condition for the uniformity of the whole correlation
+        # matrix.
+        radius_cube_frequencies = np.array(
+            np.histogram(
+                a=[polar_tuple[0] ** 3 for polar_tuple in polar_tuples],
+                bins=4,
+                range=(0, 0.5),
+            )[0],
+            dtype=float,
+        )
+        radius_cube_frequencies /= sum(radius_cube_frequencies)
+        np.testing.assert_almost_equal(radius_cube_frequencies, [1 / 4] * 4, decimal=1)
+
+        # 2. Test the uniformity of the phi angle.  Another necessary condition for the
+        # uniformity of the whole correlation matrix is that phi is uniform in [-pi, pi].
```

Comment:
Due to phi = arctan(./.), should be phi uniform in [-pi/2. pi/2]?

Reply:
You are right, and that's why I prefer `arctan2`.  `arctan2(x, y)` recognizes the signs of x, y and finds an angle in [0, 2 \pi).   Thinking about `phi` as the angle in the 2d polar coordinates, its domain is [0, 2\pi).